File: sociabletestsruby/sociable-step1/lib/file_wrapper.rb
class FileWrapper 
  def self.create data_source 
    raise ArgumentError "Pathname to file is required" if data_source == nil
    self.new(RealReader.new data_source)
  end
  def self.createNull data_source 
    raise ArgumentError "Array of fake records is required" if data_source == nil
    self.new(StubbedReader.new data_source) 
  end    
  def initialize reader
    @reader = reader
  end
  def readline 
    @reader.readline
  end      
  class StubbedReader
    def initialize data_source
      @record_list = data_source
      @next_record = 0
    end  
    def readline
      rec = @record_list[@next_record]
      raise EOFError if rec == nil
      @next_record += 1
      rec 
    end  
  end   
  class RealReader
    def initialize data_source 
      @file = File.open(data_source, "r")
    end  
    def readline
      @file.readline 
    end  
  end  
end 

File: iteration1-sociabletestsjava/src/main/java/com/neopragma/sociable/v5/FileWrapper.java
package com.neopragma.sociable.v5;

import java.io.*;

public class FileWrapper {
    private ReaderWrapper reader;
    private FileWrapper() {}
    public static FileWrapper create(String path) {
        FileWrapper tempFileSystem = new FileWrapper();
        tempFileSystem.reader = new RealReader().withPath(path);
        return tempFileSystem;
    }
    public static FileWrapper createNull(String[] inputRecords) {
        FileWrapper tempFileSystem = new FileWrapper();
        tempFileSystem.reader = new StubbedReader().withInputRecords(inputRecords);
        return tempFileSystem;
    }
    public String nextRecord() {
        return reader.nextRecord();
    }
    private interface ReaderWrapper {
        ReaderWrapper withPath(String path);
        ReaderWrapper withInputRecords(String[] inputRecords);
        String nextRecord();
    }
    private static class RealReader implements ReaderWrapper {
        private BufferedReader reader;
        public RealReader() {}
        @Override
        public ReaderWrapper withPath(String path) {
            try {
                this.reader = new BufferedReader(new FileReader(path));
            } catch (FileNotFoundException e) {
                throw new RuntimeException(e);
            }
            return this;
        }
        @Override
        public ReaderWrapper withInputRecords(String[] inputRecords) {
            return this;
        }
        @Override
        public String nextRecord() {
            try {
                return reader.readLine();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }
    private static class StubbedReader
            extends BufferedReader
            implements ReaderWrapper {
        private String[] inputRecords = null;
        private int recordIndex = 0;
        private final int EOF = -1;

        public StubbedReader() {
            super(null);
        }

        @Override
        public ReaderWrapper withPath(String path) {
            return this;
        }
        @Override
        public ReaderWrapper withInputRecords(String[] inputRecords) {
            this.inputRecords = inputRecords;
            return this;
        }
        @Override
        public String nextRecord() {
            if (inputRecords != null) {
                if (recordIndex >= inputRecords.length) {
                    return null;
                } else {
                    return inputRecords[recordIndex++];
                }
            }
            return null;
        }
    }
}
